### ISSUE TITLE
Training is slow as soon as I introduce mpi

### ISSUE THREAD

**Issue by rakshithvasudev** [horovod/horovod#1070]
**Environment:**
1. Framework: (TensorFlow, Keras, PyTorch, MXNet) : Tensorflow -gpu
2. Framework version: 1.12.0
3. Horovod version: 0.16.1
4. MPI version: 4.0.0 with cuda and ucx support
5. CUDA version: 10.0.130
6. NCCL version: 2.4.2
7. Python version: 3.6.8
8. OS and version: RHEL 7

**Checklist:**
1. Did you search issues to find if somebody asked this question before? Yes
2. If your question is about hang, did you read [this doc](https://github.com/horovod/horovod/blob/master/docs/running.md)? Yes
3. If your question is about docker, did you read [this doc](https://github.com/horovod/horovod/blob/master/docs/docker.md)? Yes

**Your question:**

Training for an epoch is faster about 350ish seconds without mpi+horovod, but as soon as I introduce mpi+horovod , training becomes slower, as high as 1179s. I wonder why this is happening. 
[timeline-1epoch.zip](https://github.com/horovod/horovod/files/3173919/timeline-1epoch.zip)

Any pointers would be appreciated.

**Comment by abditag2** [comment]
Can you clarify what is your setup? Are you comparing a single node training with and without Horovod? Also, the timeline you have attached does not show any activity beyond the first 1 second.

**Comment by rakshithvasudev** [comment]
Thanks for responding. I reran to get actual numbers again with 1 gpu and I've attached the new timeline for 1 gpu, Thanks for pointing out on the timeline. I'm using V100's and the training model is close to a resnet50 (pretrained resnet50 with some more layers). At this point, yes, I'm trying to compare with and without horovod on a single gpu. (I wouldn't assume hvd to be faster on single gpu, however it must not be ~3.5X slower either, right?). Also, I did run the same code with 4 gpus using hvd. I've included those numbers too. 

===================================================================
When I say with horovod here is what I mean: 
tf.keras Resnet50 model + hvd - 1 gpu 
Here's how I run with the above mentioned env on slurm: 

mpirun -np 1 -x NCCL_DEBUG=INFO python $HOME/resnet.py  --batch_size=128  --epochs=1
It takes 1358.88s to train the first epoch with hvd using 1 gpu.

===================================================================
tf.keras Resnet50 model  + hvd - 4 gpus 

mpirun -np 4 -x NCCL_DEBUG=INFO python $HOME/resnet.py  --batch_size=128  --epochs=1
It takes 232.58s to train the first epoch using 4 gpus.

==================================================================
Without horovod:
tf.keras Resnet 50 - 1 gpu 

python $HOME/resnet_raw.py  --batch_size=128  --epochs=1
It takes 384.96s to train the first epoch without hvd.
=================================================================== 

At this point, the effective difference using 4 gpus with hvd and 1 gpu without hvd is just 152.38s.
Like I said, I don't know what is wrong with my setup that makes it slow as soon as I introduce horovod.

[timeline-third-attempt.zip](https://github.com/horovod/horovod/files/3175172/timeline-third-attempt.zip)

**Comment by alsrgv** [comment]
@rakshithvasudev, this is suspicious because your timeline only has Broadcast operations and no Allreduce operations.  Can you share the diff between resnet_raw.py and resnet.py?

**Comment by alsrgv** [comment]
One culprit for the slowdown is CPU binding that MPI does by default.  Can you try replacing `mpirun` with `horovodrun` which passes appropriate parameters to MPI?

**Comment by rakshithvasudev** [comment]
Here's what the differences would be: 
===================================================================
resnet_raw.py:
* No horovod, so it'd be something like this :
 - Build the model (Add layers to the pretrained resnet)
 -  Add Adam optimizer 
 - Train the model    
===================================================================
resnet.py:
* This is with horovod, using tf.keras -  something like this:
- import horovod.tensorflow.keras as hvd
- Initialize horovod
- Pin GPU to be used to process local rank (one GPU per process)
- set the keras backend with tf config proto's session - something like this :

    session = tensorflow.Session(config=config)
    K.set_session(session)

- Build the model (Add layers to the pretrained resnet)
- Add Horovod Distributed Optimizer
- Add BroadcastGlobalVariablesHook and other callbacks like MetricAverageCallback(), etc
- Train the model
===================================================================
At this point, I'm not checkpointing on either versions. 

By the way, I had actually tried doing something like this : 
mpirun -np 4 \
 - bind-to none -map-by slot \
 - x NCCL_DEBUG=INFO -x LD_LIBRARY_PATH \
 -   python resnet.py

I didn't see any performance improvement. Also, when I tried to run horovodrun I encountered this error : https://github.com/horovod/horovod/issues/948

I pip uninstalled horovod and installed horovod like so:
HOROVOD_NCCL_HOME=$HOME/nccl_2.4.2-1+cuda10.0_x86_64/ HOROVOD_GPU_ALLREDUCE=NCCL pip install --upgrade-strategy only-if-needed --no-cache-dir git+https://github.com/horovod/horovod

As of now, it looks like horovodrun is a command that could be executed on my bash version. I'll keep you updated on how the performance numbers look like when the job has run successfully. I'm waiting on resources.

**Comment by rakshithvasudev** [comment]
I have an update on numbers, looks like horovodrun improved my single gpu performance. It went from 1358.88s to 403.32s, which is nice. 

However, if I scale to 4 gpus, the performance becomes bad, it is 316.28s. Just so you know what is up with the communication, here is a line from NCCL INFO :  [1,3]<stdout>:gpu028:31229:31250 [3] NCCL INFO Ring 11 : 3[3] -> 2[2] via P2P/IPC
****

I've included my slurm submission script (.sh file), source code I'm referring(.py files), output generated by 4 gpus(.o file) in the attached .zip file.  

Any thoughts and feedback would be appreciated.
 
[gh_issue.zip](https://github.com/horovod/horovod/files/3179965/gh_issue.zip)

**Comment by alsrgv** [comment]
@rakshithvasudev, thanks for the detailed information.  This is super helpful.  I looked into the timelines, and I see a pretty big gap between the time gradient allreduce of one iteration is completed and start of the next iteration.  It could be caused by the forward, or stalls in the data pipeline.

During the run, what kind of CPU utilization are you seeing (e.g. from `top`)?  How many CPU cores does this server have?  If this is caused by over-utilization of CPU due to thread starvation (too many threads), you can try reducing `--parallel_calls` to `10` with Horovod since you are going to run 4 processes on the server.

**Comment by rakshithvasudev** [comment]
Thanks for the inputs @alsrgv . It was more about fixing my tf data pipeline. I also reduced the number of parallel calls, that way I'm not oversubscribing. The server by the way has 40 cores. 

The solution for me was to follow the tf pipeline format established by tensorflow official models.

For anybody looking at this later, when I straight up implemented the way the pipeline is, I got an error that looks similar to this https://github.com/horovod/horovod/issues/72 as I started scaling to multi nodes,  the only difference about that error and my error is, mine is a bare metal run, not on K8s.

All I had to do, was to run my job like so:  

`mpirun -np 16 -bind-to none -map-by slot -x NCCL_DEBUG=INFO python official_resnet_tf_1.12.0.py  --batch_size=128  --epochs=10 
`

and now it works fine. Scales well. It is faster. I appreciate the support and feedback once again.
